src/controllers/krate/search.rs

```diff
@@ -356,4 +245,174 @@ pub async fn search(app: AppState, req: Parts) -> AppResult<Json<Value>> {
     .await
 }
 
+#[derive(Default)]
+struct FilterParams<'a> {
+    q_string: Option<&'a str>,
+    include_yanked: bool,
+    category: Option<&'a str>,
+    all_keywords: Option<&'a str>,
+    keyword: Option<&'a str>,
+    letter: Option<&'a str>,
+    user_id: Option<i32>,
+    team_id: Option<i32>,
+    following: bool,
+    has_ids: bool,
+    _auth_user_id: OnceCell<i32>,
+    _ids: OnceCell<Option<Vec<String>>>,
+}
+
+impl<'a> FilterParams<'a> {
+    fn ids(&self, req: &Parts) -> Option<&[String]> {
+        self._ids
+            .get_or_init(|| {
+                if self.has_ids {
+                    let query_bytes = req.uri.query().unwrap_or("").as_bytes();
+                    let v = url::form_urlencoded::parse(query_bytes)
+                        .filter(|(key, _)| key == "ids[]")
+                        .map(|(_, value)| value.to_string())
+                        .collect::<Vec<_>>();
+                    Some(v)
+                } else {
+                    None
+                }
+            })
+            .as_deref()
+    }
+
+    fn authed_user_id(&self, req: &Parts, conn: &mut PgConnection) -> AppResult<&i32> {
+        self._auth_user_id.get_or_try_init(|| {
+            let user_id = AuthCheck::default().check(req, conn)?.user_id();
+            Ok(user_id)
+        })
+    }
+
+    fn supports_seek(&self) -> bool {
+        // Calculating the total number of results with filters is supported but paging is not supported yet.
```

Comment:
I'm correct in thinking this isn't a regression, right, because we don't support this today?

Reply:
Yes, and I also like to make a follow up PR which expands the scope of the seek-based pagination for crates endpoint.

Reply:
For offset-based pagination, the response will remain unchanged. The only modification pertains to seek-based pagination. Previously, we provided an approximate total count; however, with this PR, the total count will align precisely with the offset-based count.